### DISCUSSION TITLE
How do we define when the schema should be updated?

### DISCUSSION THREAD

**Discussion by nickkaczmarek** [wwt/WorkflowSchema#15] [Q&A] [answered]
Was talking to @Tyler-Keith-Thompson about some things left to do on this project and he brought up schema versioning. Specifically this was the gist of his comment:

> Okay, so the thing that's missing is guidance on when the schema version has to change and what the consequences are (every library must update)

I have a good idea of _how_ to change the schema (via PR labels), but I don't have a great idea of when we should do that. For instance, what constitutes a `patch` vs `minor` vs `major`.

If I had to guess I would say that changing the type or name of a property would be a `major` change. I think `patch` versions would be basically everything else and beyond that I'm less sure. I think the below reference helps with understanding `major` and `patch` updates, but less so on the `minor` front. Opening this up to see if anyone has a good idea of the direction here.

### Keep in mind this is going to be used to define our versioning docs in the `CONTRIBUTING.md` file

[For reference, this is the guidance from semver.org](https://semver.org/#spec-item-6)

<b>Patch version Z (x.y.Z | x > 0)</b>
- MUST be incremented if only backwards compatible bug fixes are introduced. 
- A bug fix is defined as an internal change that fixes incorrect behavior.

<b>Minor version Y (x.Y.z | x > 0)</b>
- MUST be incremented if new, backwards compatible functionality is introduced to the public API. 
- It MUST be incremented if any public API functionality is marked as deprecated. 
- It MAY be incremented if substantial new functionality or improvements are introduced within the private code. 
- It MAY include patch level changes. 
- Patch version MUST be reset to 0 when minor version is incremented.

<b>Major version X (X.y.z | X > 0)</b>
- MUST be incremented if any backwards incompatible changes are introduced to the public API. 
- It MAY also include minor and patch level changes. 
- Patch and minor versions MUST be reset to 0 when major version is incremented.

**Comment by Tyler-Keith-Thompson** [comment, marked as answer]
So my take is perhaps a little strong, but I'll throw it out anyway. I don't think there's such a thing as a backward-compatible i.e. non-breaking change. The reason why is the *library*, not the consumers of the library, is doing the parsing. So if I add a field that's a non-breaking change to the agreed-upon contract, but it is a breaking change because the library has to parse the new field for it to matter and consumers have to update the library to get that parser change.

I think the version does not change:
- When the changes are metadata changes, so examples are updated, descriptions are revamped, that kind of thing.

The patch version changes:
- When the effect is additive. So there's a new field to deal with.

The minor version changes:
- When there is a break to an existing part of the contract. If we changed the object representation of `FlowRepresentableName`, for example.

The major version changes:
- When there's a fundamental shift in design, or design is solidified. An example will be we rev the major version when the schema quits changing and we are ready to release. Future changes will be when we change our very notions of how this works, maybe `sequence` goes away and you describe workflows in some new, fancier way.

**Reply by Tyler-Keith-Thompson** [reply]
NOTE: Current parser implementation statically checks the full schema version, so a patch version results in the parser throwing an error because it was not designed to parse the schema version provided. This is because identifying what constitutes a breaking change is close to impossible. Perhaps Android adds a thing that doesn't break iOS but breaks Android, or web, or something else. 

This is why my statement is contrary to perhaps popular belief on how semver works or how parsers should just try to be future-proof by trusting version numbers and meanings of major/minor/patch

**Reply by nickkaczmarek** [reply]
This works for me. I'm going to add this to `CONTRIBUTING.md` and reword the major section a little bit.

**Reply by nickkaczmarek** [reply]
I was thinking of something like this but figured I could get some quick feedback before making the pr. I tried to keep it short and simple, but not sure if I explained it well.

<hr />

## Updating the schema
If your contribution meets any of the below guidelines it's required that you add the appropriate label on the pull request.
Using these labels will run a GitHub action that revs the version accordingly.


#### Important note
Updating the version of the schema requires that all the libraries using the schema are updated to reflect the new version.
This includes [SwiftCurrent](https://github.com/wwt/SwiftCurrent).

### No label needed
- metadata changes. `ex. examples are updated, descriptions are revamped, etc.`

### Patch label
- additive changes `ex. new fields`

### Minor label
- break to the existing api. `ex. changing the way that FlowRepresentableName works`

### Major label
- fundamental shift in design. `ex. [sequence] goes away and we decide on a new way to describe workflows`
